### PR TITLE
[DeviceMemory] Add Basic Statistics to Device Memory in OpenReg

### PR DESCRIPTION
**Pull request by licy666**
Implement a complete OpenRegDeviceAllocator with the following enhancements:

- Implement memory statistics tracking (allocated/reserved bytes, allocation count)
   - Track allocation sizes for accurate memory statistics
- Refactor DeviceAllocator's inheritance relationship from c10::DeviceAllocator
  - This change is for further improvement of adding a memory caching function to DeviceMemory

Add comprehensive test coverage:
- Memory allocation/deallocation tests with statistics validation
- Storage operations and tensor-from-blob tests
- Multithreading safety tests for concurrent allocations
- Gradient tracking and requires_grad compatibility tests

Fixes #166157

### CODE REVIEW

**Comment by licy666** [comment]
@fffrog please approve for running the CI.

**Comment by fffrog** [comment]
Sorry, I've been very busy with work lately, but I'll soon have plenty of time to contribute to the community.

> @fffrog should I wait for you to take a look at this one before looking myself?

@albanD, I'll take a look at this one today. If you have any free time, please also review it. 
Thank you.

File: test/cpp_extensions/open_registration_extension/torch_openreg/csrc/runtime/OpenRegDeviceAllocator.cpp [hunk 1]
```diff
@@ -1,8 +1,205 @@
...
+
 namespace c10::openreg {
 
-static OpenRegDeviceAllocator global_openreg_alloc;
-REGISTER_ALLOCATOR(c10::DeviceType::PrivateUse1, &global_openreg_alloc);
+// Anonymous namespace for internal implementation
+namespace {
+
+// Static global allocator instance
+static OpenRegDeviceAllocator allocator;
```
Comment (fffrog) [inline test/cpp_extensions/open_registration_extension/torch_openreg/csrc/runtime/OpenRegDeviceAllocator.cpp]:
```suggestion
OpenRegDeviceAllocator g_allocator;
```

- We do not need to add static prefix in anonymous namespace, right?
- It's best to add the prefix `g_` before global variables to indicate that they are global variables.

File: test/cpp_extensions/open_registration_extension/torch_openreg/csrc/runtime/OpenRegDeviceAllocator.cpp [hunk 2]
```diff
@@ -1,8 +1,205 @@
...
+
+// Static deleter function
+static void deleteOpenRegMemory(void* ptr) {
```
Comment (fffrog) [inline test/cpp_extensions/open_registration_extension/torch_openreg/csrc/runtime/OpenRegDeviceAllocator.cpp]:
```suggestion
void deleteOpenRegMemory(void* ptr) {
```

File: test/cpp_extensions/open_registration_extension/torch_openreg/csrc/runtime/OpenRegDeviceAllocator.h [hunk 3]
```diff
@@ -1,43 +1,49 @@
...
-    return &ReportAndDelete;
-  }
-
-  void copy_data(void* dest, const void* src, std::size_t count) const final {
-    orMemcpy(dest, src, count, orMemcpyDeviceToDevice);
-  }
+
+class OpenRegDeviceAllocator final : public c10::DeviceAllocator {
+ public:
+  OpenRegDeviceAllocator();
```
Comment (fffrog) [inline test/cpp_extensions/open_registration_extension/torch_openreg/csrc/runtime/OpenRegDeviceAllocator.h:47]:
```suggestion
  OpenRegDeviceAllocator() = default;
```

File: test/cpp_extensions/open_registration_extension/torch_openreg/csrc/runtime/OpenRegDeviceAllocator.cpp [hunk 4]
```diff
@@ -1,8 +1,205 @@
...
+  allocator.freeMemory(ptr);
+}
+
+} // anonymous namespace
+
+OpenRegDeviceAllocator::OpenRegDeviceAllocator() {
+  // Initialize stats to zero (default constructor already does this)
+}
```
Comment (fffrog) [inline test/cpp_extensions/open_registration_extension/torch_openreg/csrc/runtime/OpenRegDeviceAllocator.cpp]: - We can use the `default` keyword to indicate compiler to generate default constructor for use.

```C++
class OpenRegDeviceAllocator final : public c10::DeviceAllocator {
 public:
  OpenRegDeviceAllocator() = default;
  ...
}
```

File: test/cpp_extensions/open_registration_extension/torch_openreg/csrc/runtime/OpenRegDeviceAllocator.cpp [hunk 5]
```diff
@@ -1,8 +1,205 @@
...
+    ret = orMalloc(&data, nbytes);
+
+    TORCH_CHECK(
+        ret == orSuccess && data != nullptr,
+        "Failed to allocate ", nbytes, " bytes on openreg device. ",
+        "Allocated: ", stats_.allocated_bytes[0].current, " bytes, ",
+        "Reserved: ", stats_.reserved_bytes[0].current, " bytes");
+
+    // Track allocation size for proper deallocation statistics
+    allocation_sizes_[data] = nbytes;
```
Comment (fffrog) [inline test/cpp_extensions/open_registration_extension/torch_openreg/csrc/runtime/OpenRegDeviceAllocator.cpp]: Hi, it would be even better if we could perform statistics for each device individually, instead of aggregating the memory information for all devices. Additionally, there are APIs like `getDeviceStats` and `resetPeakStats`, which both take a `device` parameter and can perform statistics for each device.

File: test/cpp_extensions/open_registration_extension/torch_openreg/csrc/runtime/OpenRegDeviceAllocator.cpp [hunk 6]
```diff
@@ -1,8 +1,205 @@
...
+bool OpenRegDeviceAllocator::initialized() {
+  // OpenReg is always initialized (simplified implementation)
+  // TODO: When caching is implemented, release all free blocks here
+  return true;
+}
+
+void OpenRegDeviceAllocator::emptyCache(MempoolId_t mempool_id) {
+  // Phase 1: Empty implementation (no caching mechanism yet)
+  // TODO: When caching is implemented, release all free blocks here
+}
```
Comment (fffrog) [inline test/cpp_extensions/open_registration_extension/torch_openreg/csrc/runtime/OpenRegDeviceAllocator.cpp]: please delete this. 
We can add it back later when we add caching support for OpenReg.

File: test/cpp_extensions/open_registration_extension/torch_openreg/csrc/runtime/OpenRegDeviceAllocator.cpp [hunk 7]
```diff
@@ -1,8 +1,205 @@
...
+
+void OpenRegDeviceAllocator::recordStream(
+    const DataPtr& ptr,
+    c10::Stream stream) {
+  // Phase 1: Empty implementation (no async stream support yet)
+  // TODO: When stream support is implemented, track which streams use this pointer
+}
```
Comment (fffrog) [inline test/cpp_extensions/open_registration_extension/torch_openreg/csrc/runtime/OpenRegDeviceAllocator.cpp]: ditto

File: test/cpp_extensions/open_registration_extension/torch_openreg/csrc/runtime/OpenRegDeviceAllocator.cpp [hunk 8]
```diff
@@ -1,8 +1,205 @@
...
+
+c10::CachingDeviceAllocator::DeviceStats OpenRegDeviceAllocator::getDeviceStats(
+    c10::DeviceIndex device) {
+  std::lock_guard<std::mutex> lock(mutex_);
+  return stats_;
+}
+
+void OpenRegDeviceAllocator::resetAccumulatedStats(c10::DeviceIndex device) {
```
Comment (fffrog) [inline test/cpp_extensions/open_registration_extension/torch_openreg/csrc/runtime/OpenRegDeviceAllocator.cpp]: We need to use the `device` parameter to return memory statistics for a specific device.

File: test/cpp_extensions/open_registration_extension/torch_openreg/csrc/runtime/OpenRegDeviceAllocator.h [hunk 9]
```diff
@@ -1,43 +1,49 @@
...
+  void resetAccumulatedStats(c10::DeviceIndex device) override;
+  void resetPeakStats(c10::DeviceIndex device) override;
+
+  // ============ Public methods for memory management ============
+
+  void freeMemory(void* ptr);
+
+ private:
+  // Device statistics
+  c10::CachingDeviceAllocator::DeviceStats stats_;
```
Comment (fffrog) [inline test/cpp_extensions/open_registration_extension/torch_openreg/csrc/runtime/OpenRegDeviceAllocator.h]: We can create a new class to represent memory management for per devices individually, and use `Delegate Pattern` to delegate detailed memory management of every device into this class.

**Comment by licy666** [comment]
@fffrog The defects have been fixed. Please take a look. Thank you!

File: test/cpp_extensions/open_registration_extension/torch_openreg/csrc/runtime/OpenRegDeviceAllocator.h [hunk 10]
```diff
@@ -1,43 +1,110 @@
...
+#include <memory>
+#include <mutex>
+#include <unordered_map>
+#include <vector>
+
 namespace c10::openreg {
-struct OpenRegDeviceAllocator final : at::Allocator {
+
+// Forward declaration
+class DeviceMemoryAllocator;
```
Comment (fffrog) [inline test/cpp_extensions/open_registration_extension/torch_openreg/csrc/runtime/OpenRegDeviceAllocator.h]: Please place the defination of `DeviceMemoryAllocator` to the top of file, so we do not need the forward declaration.

File: test/cpp_extensions/open_registration_extension/torch_openreg/csrc/runtime/OpenRegDeviceAllocator.h [hunk 11]
```diff
@@ -1,43 +1,110 @@
...
+
+  at::DataPtr allocate(size_t nbytes) override;
+  at::DeleterFnPtr raw_deleter() const override;
+  void copy_data(void* dest, const void* src, std::size_t count) const final;
+
+  // ============ Inherited from c10::DeviceAllocator ============
+
+  bool initialized() override;
+  void emptyCache(MempoolId_t mempool_id = {0, 0}) override;  // 新增
+  void recordStream(const DataPtr& ptr, c10::Stream stream) override;  // 新增
```
Comment (fffrog) [inline test/cpp_extensions/open_registration_extension/torch_openreg/csrc/runtime/OpenRegDeviceAllocator.h]: We can focus solely on supporting memory statistics, so other features (such as `emptyCache`) can be temporarily removed, which also benefits the maintainers' review.

File: test/cpp_extensions/open_registration_extension/torch_openreg/csrc/runtime/OpenRegDeviceAllocator.h [hunk 12]
```diff
@@ -1,43 +1,110 @@
...
+   * This method is idempotent and can be called multiple times.
+   */
+  void init(c10::DeviceIndex device_count);
+
+ private:
+  // Ensure device index is valid
+  void assertValidDevice(c10::DeviceIndex device);
+
+  // Get the device index for a given pointer
+  c10::DeviceIndex getDeviceForPointer(void* ptr);
```
Comment (fffrog) [inline test/cpp_extensions/open_registration_extension/torch_openreg/csrc/runtime/OpenRegDeviceAllocator.h]: Remove it please.

File: test/cpp_extensions/open_registration_extension/torch_openreg/csrc/runtime/OpenRegDeviceAllocator.cpp [hunk 13]
```diff
@@ -1,8 +1,336 @@
...
+  auto ret = orGetDevice(&current_device_index);
+  TORCH_CHECK(ret == orSuccess, "Failed to get current OpenReg device");
+
+  // Auto-initialize on first use
+  if (device_allocators_.empty()) {
+    std::lock_guard<std::recursive_mutex> lock(mutex_);
+    if (device_allocators_.empty()) { // double-check
+      init(c10::openreg::device_count());
+    }
+  }
```
Comment (fffrog) [inline test/cpp_extensions/open_registration_extension/torch_openreg/csrc/runtime/OpenRegDeviceAllocator.cpp]: We will ensure that the memory is initialized during device initialization.

File: test/cpp_extensions/open_registration_extension/torch_openreg/csrc/runtime/OpenRegDeviceAllocator.cpp [hunk 14]
```diff
@@ -1,8 +1,336 @@
...
+  // OpenReg doesn't implement caching yet
+  // TODO: When caching is implemented, release all free blocks here
+}
+
+void OpenRegDeviceAllocator::recordStream(
+    const DataPtr& ptr,
+    c10::Stream stream) {
+  // OpenReg doesn't track stream usage yet
+  // TODO: When stream support is added, track which streams are using this pointer
+}
```
Comment (fffrog) [inline test/cpp_extensions/open_registration_extension/torch_openreg/csrc/runtime/OpenRegDeviceAllocator.cpp:270]: Remove it please

File: test/cpp_extensions/open_registration_extension/torch_openreg/csrc/runtime/OpenRegDeviceAllocator.h (hunk 10 above)
Comment (licy666) [reply test/cpp_extensions/open_registration_extension/torch_openreg/csrc/runtime/OpenRegDeviceAllocator.h]: done.

File: test/cpp_extensions/open_registration_extension/torch_openreg/csrc/runtime/OpenRegDeviceAllocator.h (hunk 12 above)
Comment (licy666) [reply test/cpp_extensions/open_registration_extension/torch_openreg/csrc/runtime/OpenRegDeviceAllocator.h]: done

File: test/cpp_extensions/open_registration_extension/torch_openreg/csrc/runtime/OpenRegDeviceAllocator.cpp [hunk 15]
```diff
@@ -1,8 +1,281 @@
...
+
+  auto ret = orFree(ptr);
+
+  if (ret == orSuccess) {
+    auto it = allocation_sizes_.find(ptr);
+    if (it != allocation_sizes_.end()) {
+      size_t nbytes = it->second;
+
+      using namespace c10::CachingAllocator;
+      constexpr size_t kAggregate = static_cast<size_t>(StatType::AGGREGATE);
```
Comment (fffrog) [inline test/cpp_extensions/open_registration_extension/torch_openreg/csrc/runtime/OpenRegDeviceAllocator.cpp]: Move this two line to the top of the file.

File: test/cpp_extensions/open_registration_extension/torch_openreg/csrc/runtime/OpenRegDeviceAllocator.cpp [hunk 16]
```diff
@@ -1,8 +1,281 @@
...
+
+c10::CachingDeviceAllocator::DeviceStats DeviceMemoryAllocator::getStats() {
+  std::lock_guard<std::recursive_mutex> lock(mutex_);
+  return stats_;
+}
+
+void DeviceMemoryAllocator::resetAccumulatedStats() {
+  std::lock_guard<std::recursive_mutex> lock(mutex_);
+
+  using namespace c10::CachingAllocator;
```
Comment (fffrog) [inline test/cpp_extensions/open_registration_extension/torch_openreg/csrc/runtime/OpenRegDeviceAllocator.cpp]: Move this line to top level because this have been used in many place.

File: test/cpp_extensions/open_registration_extension/torch_openreg/csrc/runtime/OpenRegDeviceAllocator.cpp [hunk 17]
```diff
@@ -1,8 +1,281 @@
...
+  stats_.num_ooms = 0;
+  stats_.num_sync_all_streams = 0;
+  stats_.num_device_alloc = 0;
+  stats_.num_device_free = 0;
+}
+
+void DeviceMemoryAllocator::resetPeakStats() {
+  std::lock_guard<std::recursive_mutex> lock(mutex_);
+
+  using namespace c10::CachingAllocator;
```
Comment (fffrog) [inline test/cpp_extensions/open_registration_extension/torch_openreg/csrc/runtime/OpenRegDeviceAllocator.cpp]: ditto

File: test/cpp_extensions/open_registration_extension/torch_openreg/csrc/runtime/OpenRegDeviceAllocator.cpp [hunk 18]
```diff
@@ -1,8 +1,281 @@
...
+      "Reserved: ",
+      stats_.reserved_bytes[0].current,
+      " bytes");
+
+  // Track allocation size for proper deallocation statistics
+  allocation_sizes_[data] = nbytes;
+
+  // Update statistics
+  using namespace c10::CachingAllocator;
+  constexpr size_t kAggregate = static_cast<size_t>(StatType::AGGREGATE);
```
Comment (fffrog) [inline test/cpp_extensions/open_registration_extension/torch_openreg/csrc/runtime/OpenRegDeviceAllocator.cpp]: ditto

**Review by fffrog — APPROVED** [review]
LGTM, thank you a lot.

File: test/cpp_extensions/open_registration_extension/torch_openreg/csrc/runtime/OpenRegDeviceAllocator.cpp [hunk 19]
```diff
@@ -1,8 +1,275 @@
...
+using namespace c10::CachingAllocator;
 
 namespace c10::openreg {
 
-static OpenRegDeviceAllocator global_openreg_alloc;
-REGISTER_ALLOCATOR(c10::DeviceType::PrivateUse1, &global_openreg_alloc);
+constexpr size_t kAggregate = static_cast<size_t>(StatType::AGGREGATE);
+
+
+DeviceMemoryAllocator::DeviceMemoryAllocator(c10::DeviceIndex device_index)
```
Comment (fffrog) [inline test/cpp_extensions/open_registration_extension/torch_openreg/csrc/runtime/OpenRegDeviceAllocator.cpp:14]: I prefer to move the simple and concise implementation to the header file (this can be done in the next PR).

File: test/cpp_extensions/open_registration_extension/torch_openreg/tests/test_memory.py [hunk 20]
```diff
@@ -1,9 +1,404 @@
...
+    #    """Test allocation on different device indices if supported."""
+    # Test device 0
+    #    x = torch.empty(100, device="openreg:0")
+    #    self.assertEqual(x.device.type, "openreg")
+    #    self.assertEqual(x.device.index, 0)
+
+    # Test cross device
+    #    y = torch.empty(100, device="openreg:1")
+    #    self.assertEqual(y.device.type, "openreg")
+    #    self.assertEqual(y.device.index, 1)
```
Comment (fffrog) [inline test/cpp_extensions/open_registration_extension/torch_openreg/tests/test_memory.py]: Hey, we can safely remove this one testcases.

**Comment by licy666** [comment]
> @pytorchbot started a rebase job onto [refs/remotes/origin/main](https://github.com/pytorch/pytorch/commit/4de24bcc569ec19b20756e3aa01f874bc435555d). Check the current status [here](https://github.com/pytorch/pytorch/actions/runs/19334127012)

done
